packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,370 @@
+    required this.selection,
+    required this.composing,
+  }) : assert(oldText != null),
+       assert(selection != null),
+       assert(composing != null);
+
+  /// Creates an instance of this class from a JSON object by inferring the
+  /// type of delta based on values sent from the engine.
+  factory TextEditingDelta.fromJSON(Map<String, dynamic> encoded) {
+    // An insertion delta is one where replacement destination is collapsed.
+    //
+    // A deletion delta is one where the replacement source is empty.
+    //
+    // An insertion/deletion can still occur when the replacement destination is not
+    // collapsed, or the replacement source is not empty.
+    //
+    // On native platforms when composing text, the entire composing region is
+    // replaced on input, rather than reporting character by character
+    // insertion/deletion. In these cases we can detect if there was an
+    // insertion/deletion by checking if the text inside the original composing
+    // region was modified by the replacement. If the text is the same then we have
+    // an insertion/deletion. If the text is different then we can say we have
+    // a replacement.
+    //
+    // For example say we are currently composing the word: 'world'.
+    // Our current state is 'worl|' with the cursor at the end of 'l'. If we
+    // input the character 'd', the platform will tell us 'worl' was replaced
+    // with 'world' at range (0,4). Here we can check if the text found in the
+    // composing region (0,4) has been modified. We see that it hasn't because
+    // 'worl' == 'worl', so this means that the text in
+    // 'world'{replacementDestinationEnd, replacementDestinationStart + replacementSourceEnd}
+    // can be considered an insertion. In this case we inserted 'd'.
+    //
+    // Similarly for a a deletion, say we are currently composing the word: 'worl'.
+    // Our current state is 'world|' with the cursor at the end of 'd'. If we
+    // press backspace to delete the character 'd', the platform will tell us 'world'
+    // was replaced with 'worl' at range (0,5). Here we can check if the text found
+    // in the new composing region, is the same as the replacement text. We can do this
+    // by using oldText{replacementDestinationStart, replacementDestinationStart + replacementSourceEnd}
+    // which in this case is 'worl'. We then compare 'worl' with 'worl' and
+    // verify that they are the same. This means that the text in
+    // 'world'{replacementDestinationEnd, replacementDestinationStart + replacementSourceEnd} was deleted.
+    // In this case the character 'd' was deleted.
+    //
+    // A replacement delta occurs when the original composing region has been
+    // modified.
+    //
+    // A non text update delta occurs when the selection and/or composing region
+    // has been changed by the platform, and there have been no changes to the
+    // text value.
+    final String oldText = encoded['oldText'] as String;
```

Comment:
Instead of parsing the type from the properties of the data, can we concretely determine this from the source in the embedding and pass this information up? This parsing seems to work, but I imagine it could be vulnerable to edge cases in the case of wild and crazy deltas. It would be better to be absolutely certain when possible, and fallback on this algo only when the source cannot distinguish for certain.

Reply:
Part of this algorithm is using the method definitions from the platforms to concretely determine if something an insertion or a deletion.

From android Editable documentation for their `replace()` method.
```
     * Replaces the specified range (<code>st&hellip;en</code>) of text in this
     * Editable with a copy of the slice <code>start&hellip;end</code> from
     * <code>source</code>.  The destination slice may be empty, in which case
     * the operation is an insertion, or the source slice may be empty,
     * in which case the operation is a deletion.
```

Similarly looking at iOS `deleteBackwards` method https://github.com/flutter/engine/blob/e8c44adc49d56e0d00a74dde5f79ddfe526fbd37/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L1468 that also uses an `empty` source string as a replacement for the deleted range. `insertText` similarly inserts some text at the current selection. Both of these methods are convenience methods for `replaceRange WithText`.

The iOS and Android PR's before the inference logic was moved to the framework where very similar.
flutter/engine@2148347
flutter/engine@402280e
The only real new case iOS was handling is that when typing with CJK keyboard (simplified chinese pinyin for example) it actually first types the english characters first and then replaces them with the chinese characters. On Android the word is not inserted into the editing state until it has been fully composed. Keeping in mind there similarity I thought it would be better to centralize all the logic in the framework. 

The web also behaves similarly but less information is given to us upfront since we delegate all of the text input control to the `<input>/<textarea>` element. In this case we must use the relevant API's like `beforeinput` and `compositionupdate` to capture the parameters we need. Since nothing is given to us upfront, we must normalize the parameters into a common format.

The only ambiguous part really comes from dealing with cases of the composing region. Where we check if the original composing region has been changed, if it has then it is a replacement, if it hasn't then it is either an insertion/deletion. That is the main inference going on.

Reply:
Ok, thanks for the breakdown, that SGTM